.github/workflows/release.yaml

```diff
@@ -0,0 +1,211 @@
+        name: Prepare release metadata
+        runs-on: ubuntu-latest
+        outputs:
+            version_number: ${{ steps.release_metadata.outputs.version_number }}
+            tag_name: ${{ steps.release_metadata.outputs.tag_name }}
+            changelog: ${{ steps.release_metadata.outputs.changelog }}
+            release_notes: ${{ steps.release_metadata.outputs.release_notes }}
+        steps:
+            -   uses: apify/workflows/git-cliff-release@main
+                name: Prepare release metadata
+                id: release_metadata
+                with:
+                    release_type: ${{ inputs.release_type }}
+                    custom_version: ${{ inputs.custom_version }}
+                    existing_changelog_path: CHANGELOG.md
+
+
+    wait_for_checks:
+        name: Wait for code checks to pass
+        runs-on: ubuntu-latest
+        steps:
+            -   uses: lewagon/wait-on-check-action@v1.3.4
+                with:
+                    ref: ${{ github.ref }}
+                    repo-token: ${{ secrets.GITHUB_TOKEN }}
+                    check-regexp: (Build & Test .*|Test Python template support|Lint|Docs build)
+                    wait-interval: 5
+
+    update_changelog:
+        needs: [ release_metadata, wait_for_checks ]
+        name: Update changelog
+        runs-on: ubuntu-latest
+        outputs:
+            changelog_commitish: ${{ steps.commit.outputs.commit_long_sha || github.sha }}
+
+        steps:
+            -   name: Checkout repository
+                uses: actions/checkout@v4
+                with:
+                    token: ${{ secrets.APIFY_SERVICE_ACCOUNT_GITHUB_TOKEN }}
+
+            -   name: Use Node.js 22
+                uses: actions/setup-node@v4
+                with:
+                    node-version: 22
+
+            -   name: Enable corepack
+                run: |
+                    corepack enable
+                    corepack prepare yarn@stable --activate
+
+            -   name: Activate cache for Node.js 20
+                uses: actions/setup-node@v4
+                with:
+                    cache: yarn
+
+            -   name: Update package version in package.json
+                run: npm version --no-git-tag-version --allow-same-version ${{ needs.release_metadata.outputs.version_number }}
+
+            -   name: Update CHANGELOG.md
+                uses: DamianReeves/write-file-action@master
+                with:
+                    path: CHANGELOG.md
+                    write-mode: overwrite
+                    contents: ${{ needs.release_metadata.outputs.changelog }}
+
+            -   name: Commit changes
+                id: commit
+                uses: EndBug/add-and-commit@v9
+                with:
+                    author_name: Apify Release Bot
+                    author_email: noreply@apify.com
+                    message: "chore(release): Update changelog and package version [skip ci]"
+
+    create_github_release:
```

Comment:
Shouldn't this happen _after_ npm publish?

Reply:
Um, which one of the two? :smile:

Reply:
create_github_release

Reply:
I mean, it's not a problem to arrange that. I'm not sure why though :slightly_smiling_face:

Reply:
I think the problem here is that if publishing fails (e.g. because NPM registry hiccup, seen that many times), you end up with the changes in changelog and GH tags/releases you need to undo manually. It would be great to solve this, but I wouldn't care about it that much right now, we were trying to improve the old system that suffered from the very same issue, our monorepos also suffer from this.

Reply:
Hmm with this new approach I guess you can, good point, previously this was tied to other things like version bump detection (at least with the monorepo setup I recall it wasn't that simple).